### ISSUE TITLE
Excessive Zotero API requests

### ISSUE THREAD

**Issue by dstillman** [alixlahuec/zotero-roam#175]
We're seeing a huge number of requests hitting the Zotero API that seem to be coming from this plugin. Most of these are repeated requests that don't appear to be using any caching and are just repeatedly requesting users' entire libraries.

For example, on Feb. 1 we received 2.1 million requests with a `Referer` of `https://roamresearch.com/`, accessing just 132 unique libraries. Even cutting that in half to exclude the CORS preflight OPTIONS requests, that's almost 8,000 requests — or 800,000 objects — per library. Picking one user and one set of query parameters at random, we received 319 requests just for `/users/<userID>/items?start=200&limit=100`.

It's OK for a Zotero API client to download all data for a library if necessary, but all clients need to use [appropriate caching and versioning](https://www.zotero.org/support/dev/web_api/v3/basics#caching) for repeated requests, using both `If-Modified-Since-Version` and `?since=`. It should almost never be necessary to make more than one request if a library hasn't changed since the last time data was accessed. (I'm not seeing any conditional requests currently. It looks like there is some attempt to use `?since=`, but it may be broken — e.g., there were 487 requests for `/users/<userID>/items?since=0&start=0&limit=100` for the same library on Feb. 1.)

I'm not sure what kind of storage is possible in the Roam environment, but I'm assuming there's something that plugins can use to cache data.

We can give you a little bit of time to fix this, but we're going to have to start throttling these requests pretty soon.

Let us know if you have any questions. We're also always happy to answer API questions on <a href="https://groups.google.com/g/zotero-dev">zotero-dev</a>.

**Comment by alixlahuec** [comment]
Thanks for reaching out @dstillman !

For some context, Roam doesn't provide any storage/caching to plugins - right now, every time a user starts a session, the plugin requests all of the data for each of the user's libraries. For some users with 1000's of items, this means dozens of API calls just to load the initial data - not exactly scalable. (Then, regular polling for updates, which uses `since` and _some_ caching, but it all applies only within the current session)

I started working on a solution in December '22 but it's not quite ready yet - aiming to get it out this month if I can.
That should reduce API usage significantly.

The calls volume sounds suspiciously high though, so I did some investigating and there's definitely a bug with how `since` is used.
It seems to have been introduced in December, by a memoization with incorrect dependencies. Looks like the value of `since` is being memoized as `null`, which is treated as `0` by subsequent API calls - i.e the latest known version gets ignored.

That's a major bug, will fix ASAP -- and I'll bump the priority on caching as well, so it can be out sooner rather than later
